Normalize intermediate node names to better utilize cache

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149898
* __->__ #149415

This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

We do a normalization pass that effectively ensures that the name indexes monotonically increase. This typically
already happens but in some cases, such as in HOPs, the invariant could be broken without normalization. Below we
show an example where cond previously would have jumped from getitem_3 to get_item_2, but with normalization correctly uses getitem_4 after getitem_3.

We've run this on the same model internally and confirmed with change we now get a cache hit.

bobrenjc93 committed:
normalize intermediate node names

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71407299).

aorenste commented on torch/_dynamo/output_graph.py:
```diff
@@ -1765,6 +1766,24 @@ def remove_tensorify_specialized_graphargs(self) -> None:
+        base_name_counter = {}
+
+        for node in intermediate_nodes:
+            name_parts = node.name.split('_')
+            if len(name_parts) > 1 and name_parts[-1].isdigit():
+                base_name = '_'.join(name_parts[:-1])
+            else:
+                base_name = node.name
```
If you use `rsplit('_', 1)` then you can avoid the `join`...

bobrenjc93 committed:
Update on "normalize intermediate node names"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71407299](https://our.internmc.facebook.com/intern/diff/D71407299)

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71407299](https://our.internmc.facebook.com/intern/diff/D71407299)

[ghstack-poisoned]

oulgen commented on torch/_dynamo/output_graph.py:
```diff
@@ -1765,6 +1766,28 @@ def remove_tensorify_specialized_graphargs(self) -> None:
                     self.remove_node(u)
                 self.remove_node(node)
 
+    def normalize_intermediate_node_names(self) -> None:
+        intermediate_nodes = [
+            node
+            for node in self.graph.nodes
+            if node.op != "placeholder" and node.op != "output"
+        ]
+        base_name_counter = {}
+
+        for node in intermediate_nodes:
```
```suggestion
    def normalize_intermediate_node_names(self) -> None:
        base_name_counter = {}

        for node in self.graph.nodes:
            if node.op == "placeholder" and node.op == "output":
                continue
```

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

aorenste commented:
I think I'd like to see a unit test. As you point out the tests test this but I don't think they test that it actually works - just that the new names aren't "problematic".

I don't think you should need to factor anything out - just mock up a graph with some interesting cases and make sure the normalization adjusts the names to be as expected.

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "normalize intermediate node names"


This change was motivated by internal use case (https://fb.workplace.com/groups/1553867532149891/?multi_permalinks=1708481206688522&comment_id=1711739699696006&notif_id=1742399826944239&notif_t=work_feedback_reaction_generic&ref=notif) where we were producing different intermediate node names for the exact same code. This normalization pass does an alpha renaming of intermediate variables so that more isomorphic graphs now result in the same dynamo outputted graph.

Regarding testing, I re-ran the model that was cache missing with this PR patched in and now it cache hits. Since this is a normalization pass, almost every test case will exercise this code path. If reviewers insist, I can factor out this normalization pass into its own module with unit tests, but it's unclear to me if that's worth it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

aorenste commented on torch/_dynamo/output_graph.py:
```diff
@@ -1766,6 +1767,37 @@ def remove_tensorify_specialized_graphargs(self) -> None:
+            name_parts = node.name.rsplit("_", 1)
+            if len(name_parts) > 1 and name_parts[1].isdigit():
+                base_name = name_parts[0]
+                base_name_counter[base_name] = max(
+                    base_name_counter[base_name], int(name_parts[1])
+                )
```
So the thing I don't like about this is it means that a graph like
```
%getitem: abc
%getitem_12: def
%getitem_1: ghi
```
and
```
%getitem: abc
%getitem_1: def
%getitem_2: ghi
```
will normalize to different things.

Do we need the `max`? What would go wrong if we just used the next number in the sequence?

oulgen replied:
I also agree with @aorenste

bobrenjc93 replied:
IIRC the problem is if you have something like this

``
%getitem_2: def
%getitem_1: def
```

Then you run into the problem where you end up with duplicate names. I just put up a canary here to verify: https://github.com/pytorch/pytorch/pull/149896

We could bookkeep which indexes have been allocated, but that also felt heavy handed. I suppose this really stemmed from things breaking down when I normalized everything including the very first occurrence. If we always ensured we started at [op]_0, then we'd solve both problems simultaneously. I just put up another canary here https://github.com/pytorch/pytorch/pull/149898.

Depending on the scope of the blast radius of those two canaries, I do think may be worth doing something like the current PR that takes an incremental step to more normalized graphs. While not perfect, it's at least a first gradient step towards normalized graphs and better cache utilization.

aorenste replied:
I don't understand - isn't the code rewriting all the names? why wouldn't that example be rewritten as:
```
%getitem: def
%getitem_1: def
```
there can't be any existing stringy references to the names or the existing rewrite wouldn't work (you'd have dangling names).

bobrenjc93 replied:
It currently doesn't rewrite the first occurrence:

```
            else:
                base_name = node.name

            if base_name_counter[base_name] == 0:
                node.name = f"{node.name}"
```

Anyways, I took a deeper look at the failures of the canary diffs and think they are tractable (mostly HOP related). I'll try to consolidate on a simpler pass and fix up the HOPs to conform to this new scheme.

aorenste replied:
If that's the only problem it seems like it would be pretty trivial to rewrite the first instance as well...

bobrenjc93 replied:
Yep that's what I mean by "normalize all fx nodes".

aorenste commented on torch/_dynamo/output_graph.py:
```diff
@@ -1766,6 +1767,37 @@ def remove_tensorify_specialized_graphargs(self) -> None:
+            else:
+                base_name = node.name
+
+            if base_name_counter[base_name] == 0:
+                node.name = f"{node.name}"
+            else:
+                node.name = f"{base_name}_{base_name_counter[base_name]}"
```
nit:
```suggestion
            if (next_id := base_name_counter[base_name]) == 0:
                node.name = f"{node.name}"
            else:
                node.name = f"{base_name}_{next_id}"
```

bobrenjc93 commented:
I'm going to memorialize this PR by abandoning it as it is a clean run with normalization. I'll be iterating on https://github.com/pytorch/pytorch/pull/149898 moving forward.
